docs/index.bs

```diff
@@ -3288,29 +3288,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Run the following substeps [=in parallel=], but [=abort when=] |fetchController|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |preloadFetchController| to the result of [=Fetch|fetching=] |preloadRequest|.
+
+                      To [=fetch/processResponse=] for |navigationPreloadResponse|, run these substeps:
+
+                      1. If |navigationPreloadResponse|'s [=response/type=] is "`error`", reject |preloadResponse| with a `TypeError` and terminate these substeps.
+                      1. Associate |preloadResponseObject| with |navigationPreloadResponse|.
+                      1. Resolve |preloadResponse| with |preloadResponseObject|.
+              1. [=If aborted=], then:
+                  1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
+                  1. [=fetch controller/Abort=] |preloadFetchController| with |deserializedError|.
+          1. Else if |timingInfo|'s [=service worker timing info/worker matched router source=] is not "{{RouterSourceEnum/fetch-event}}", a user agent may run the following substeps:
+
+              Note: A user agent may speculatively dispatch a network request in parallel with creating a fetch event in order to minimize the bootstrap cost.
+
+              1. Let |autoPreloadFetchController| be null.
+              1. Set |responseForAutoPreload| to be a [=race response=] whose [=race response/value=] is "<code>pending</code>".
+              1. Run the following substeps [=in parallel=], but [=abort when=] |fetchController|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |autoPreloadFetchController| to the result of calling [=fetch=] given |request|, with [=fetch/processResponse=] set to the following steps given a [=/response=] |autoPreloadRequestResponse|:
+                  1. If |autoPreloadRequestResponse|'s [=response/status=] is [=ok status=], set |responseForAutoPreload|'s [=race response/value=] to |autoPreloadRequestResponse|.
+              1. [=If aborted=] and |autoPreloadFetchController| is not null, then:
+                  1. [=fetch controller/Abort=] |autoPreloadFetchController|.
+                  1. Set |responseForAutoPreload| to a [=race response=] whose [=race response/value=] is null.
+              1. Resolve |preloadResponse| with undefined.
       1. Else, resolve |preloadResponse| with undefined.
-      1. Let |fetchResult| be the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, |preloadResponse|, and null.
+      1. Let |fetchResult| be the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, |preloadResponse|, and |responseForAutoPreload|.
```

Comment:
In the current WebKit implementation, if service worker does not call respondWith, the process doing networking is notified and it will reuse the responseForAutoPreload directly.
AIUI, the mechanism as described here will likely reuse the response in the renderer process, which might a bit slower. I am not sure whether this is observable so maybe the spec is fine as is.

Reply:
I feel you assume https://w3c.github.io/ServiceWorker/#create-fetch-event-and-dispatch-algorithm is handled in the renderer process.  I understand that the Chromium implementation is actually not.  Most of Step 17 is handled in the renderer process, but other steps in the algorithm would be handled in the browser process.

Reply:
Oh, that is right, the algorithm is perfectly fine as is.